Add 'Automating Pulumi Import to Bring Manually Created Resources into IaC' (#2431)
pulumi/pulumi-hugo: themes/default/content/blog/automating-pulumi-import-to-bring-manually-created-resources-into-iac/index.md

   pulumi_resources = []

   route_tables = ec2_client.describe_route_tables()['RouteTables']

   for route_table in route_tables:
       for association in route_table["Associations"]:
           if 'SubnetId' not in association:
               continue
           pulumi_resources.append({
               "type": "aws:ec2/routeTableAssociation:RouteTableAssociation",
               "name": f"import-{association['RouteTableAssociationId']}",
               "id": f"{association['SubnetId']}/{route_table['RouteTableId']}",
           })

   return pulumi_resources
```

We can run the entire account scraper program to output our resources to a single JSON file by running the following command. Detailed instructions are contained in the account scraper repository:

```bash
python account_scraper.py > ../path/to/file.json
```

## Bulk Import

Now that we’ve generated our JSON file, we can use the `pulumi import` command to import all of our resources in a single command and output our code to its own file:

```bash
pulumi import -f ../path/to/file.json -o imported-resources.ts -y
```

And with that, our resources are now managed by Pulumi!

> ## Potential Next Steps

Reviewer: id drop potential from this

Author: Done.

---

> title: "Automating Pulumi Import with Manually Created Resources"

Reviewer 2: Something about the way this is phrased makes it sound like "manually created resources" are the feature -- like we're using manually created resources to automate Pulumi import, which doesn't quite capture it. I don't have a great alternative at the ready, but will think on it -- just wanted to give that feedback, though, since you asked specifically for it.

Author: I'm in the same boat - I am pretty happy with the content but really struggling for a fitting title.

Author: Gonna mark as resolved for now since neither of us has something clearly better.

date: 2023-01-23

> meta_desc: How to bring order to a mess of manually created cloud resources.

Reviewer 2: Maybe "order to chaos" would work a little better?

```suggestion
meta_desc: How to bring order to the chaos of manually created cloud resources.
```

authors:
    - josh-kodroff

> tags:
>     - aws

Reviewer 2: ```suggestion
tags:
    - aws
    - import
```

Introducing modern cloud infrastructure tooling and automation is relatively simple (or at least more straightforward) when organizations decide to adopt IaC from the start of their cloud journey, but what about organizations who have gone far enough down the route of manually created cloud infrastructure to see the perils of that approach? Many teams come to this realization only when they’ve deployed too many production workloads to start over from scratch. If your organization is looking at Pulumi as an IaC solution, it’s worth bringing these resources under management because of the low effort and high value of having a single pane of glass to manage all of your resources.

Bringing resources under IaC (and Pulumi specifically) gives an organization an immediate return on investment. Note that while we list some benefits specific to the Pulumi Service, Pulumi supports a number of backends for both [state management](https://www.pulumi.com/docs/intro/concepts/state/#deciding-on-a-state-backend) and [secrets management](https://www.pulumi.com/docs/intro/concepts/secrets/#available-encryption-providers) and the Pulumi Service is not required in order to adopt Pulumi:

> * **Version control/change history:** With IaC placed under version control, teams can easily roll back to previous versions of infrastructure, see who changed what, and when, which provides critical visibility into how infrastructure has evolved over time. The Pulumi Service provides a great experience here, with a dashboard that collects version history and deployment logs along with git contextual data in a “single pane of glass”.

Author: Fixed.

           })

   return pulumi_resources
```

We can run the entire account scraper program to output our resources to a single JSON file by running the following command. Detailed instructions are contained in the account scraper repository:

```bash
python account_scraper.py > ../path/to/file.json
```

## Bulk Import

Now that we’ve generated our JSON file, we can use the `pulumi import` command to import all of our resources in a single command and output our code to its own file:

```bash
pulumi import -f ../path/to/file.json -o imported-resources.ts -y
```

And with that, our resources are now managed by Pulumi!

## Potential Next Steps

Now that we have our resources under Pulumi management, we might want to consider some additional steps to build upon our IaC solution:

**Set up a CI/CD pipeline:** By adding our Pulumi code to a delivery pipeline, we can help ensure that all infrastructure changes go through an automated process. Pulumi has helpful integrations and documentation for many popular tools. See [Continuous Delivery](https://www.pulumi.com/docs/guides/continuous-delivery/) in the Pulumi docs for more information.

> **Add policy as code:** By adding policy as code to our IaC pipeline, we can ensure that our infrastructure is in compliance with any security or regulatory requirements before it’s ever provisioned in the cloud, as well as checks on any existing Pulumi resources. For more information on Pulumi’s policy as code capabilities, see [CrossGuard (Policy as Code)](https://www.pulumi.com/docs/guides/crossguard/) in the Pulumi docs.

Author: This is intended to be a bulleted list and has been fixed. Sloppy job on my part with the copy/paste from Google Docs.
